### PR TITLE
Sheland and Laura_oo-ride-share P2

### PR DESCRIPTION
**Pull request by lauracodecreations**
# OO Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | When working how to add the trips driven by each driver, we decided to use our add_driven_trip in our load trip_method, so that each trip can be added since load_trip creates a trip at a time
Describe any examples of composition that you encountered in this project, if any | each driver has a list of driven trips. Therefore, is a one-to-many relationship
Describe the relationship between `User` and `Driver`| Driver is a subclass (child) of User. The driver's id is the same as the user's id. Therefore, one-to-one 
Describe a _nominal test_ that you wrote for this assignment. | for the average_rating method, we wrote a test that verifies if the method correctly calculates the average rating.
Describe an _edge case test_ that you wrote for this assignment | for the average_rating method, we wrote a test to verify that it returns 0 when there are not trips
Describe a concept that you/your pair gained more clarity on as you worked on this assignment | inherence
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | we talked about our background in terms of coming from an ask or guess culture and how that played a role that improved over time in our communication and interactions with each other. We talked about how we learn and how to translates to us when we are thinking about how to solve a problem.

### CODE REVIEW

File: lib/trip.rb [hunk 1]
```diff
@@ -1,26 +1,58 @@
...
+      #@passenger will be updated after replacing
       @start_time = input[:start_time]
       @end_time = input[:end_time]
       @cost = input[:cost]
       @rating = input[:rating]
+      @driver_id = input[:driver_id]
+      @driver = input[:driver]
+
 
       if @rating > 5 || @rating < 1
```
Comment (CheezItMan) [inline lib/trip.rb:30]: For wave 3, you need to check for incomplete trips, so you would need something like:

```ruby
if @rating.nil? || @rating > 5 || @rating < 1
```

You also need another block to handle the fact that @end_time might be `nil`.  

```ruby
unless @end_time.nil?
  difference = @end_time - @start_time
  raise ArgumentError, 'Endtime must be after start-time' if difference < 0
end
```

File: lib/trip.rb [hunk 2]
```diff
@@ -1,26 +1,58 @@
...
     end
 
     def inspect
       "#<#{self.class.name}:0x#{self.object_id.to_s(16)} " +
       "ID=#{id.inspect} " +
       "PassengerID=#{passenger&.id.inspect}>"
     end
+
+    def duration
+      duration = @end_time - @start_time
```
Comment (CheezItMan) [inline lib/trip.rb:51]: Should also handle a potentially incomplete trip.

File: lib/driver.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      #input[:trips] = [] if input[:trips] == nil
+      #@driven_trips = input[:trips]
+      if @driven_trips.nil?
+        @driven_trips = []
+      end
+      #raise ArgumentError if not VALID_STATUS.include?(status)
+    end
+
+    def average_rating
+      ratings = @driven_trips.map do |trip|
```
Comment (CheezItMan) [inline lib/driver.rb:30]: Need to exclude incomplete trips
```ruby
ratings = @driven_trips.select {|trip| trip.rating != nil }.map do |trip|
```

File: lib/driver.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+      return average_rating
+    end
+
+    def total_revenue
+      binding.pry
+      total = 0
+      return total if @driven_trips.length == 0
+
+      @driven_trips.each do |trip|
```
Comment (CheezItMan) [inline lib/driver.rb:45]: You also need to exclude incomplete trips.

File: lib/user.rb [hunk 5]
```diff
@@ -13,8 +15,28 @@ def initialize(input)
...
     end
 
+    # this is used in the load trip method
+    # it collects the instances of trips and it is saves in the instance of trips for each user ^ anove
     def add_trip(trip)
       @trips << trip
     end
+
+    def net_expenditures
+      net_total = 0
```
Comment (CheezItMan) [inline lib/user.rb:25]: Need to handle incomplete trips

File: lib/user.rb [hunk 6]
```diff
@@ -13,8 +15,28 @@ def initialize(input)
...
+      @trips.each do |trip|
+      net_total += trip.cost
+      end
+      return net_total
+    end
+
+
+    def total_time_spent #total time spent on user's trip
+      net_time = 0
```
Comment (CheezItMan) [inline lib/user.rb:34]: Need to handle incomplete trips

File: specs/driver_spec.rb [hunk 7]
```diff
@@ -95,7 +96,48 @@
...
+   it "correctly calculates the total_revenue" do
+     @driver1 = RideShare::Driver.new(id: 54, name: "Rogers Bartell IV",
+                                     vin: "1C9EVBRM0YBC564DZ")
+
+     trip = RideShare::Trip.new(id: 8, driver: @driver1, passenger: nil,
+                               start_time: Time.parse("2016-08-08"),
+                               end_time: Time.parse("2016-08-08"),
+                               cost: 5, rating: 5)
+
+      trip2 = RideShare::Trip.new(id: 8, driver: @driver1, passenger: nil,
```
Comment (CheezItMan) [inline specs/driver_spec.rb:133]: no rating included here, causing an ArgumentError to trigger.

File: lib/driver.rb [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        total += trip.cost
+      end
+      #[12, 32, 43]
+      cost_sum = total * DRIVER_PERCENTAGE
+      total_revenue = cost_sum - FEE
+      return total_revenue
+    end
+
+
```
Comment (CheezItMan) [inline lib/driver.rb:54]: Need a `net_expenditures` method.

```ruby
def net_expenditures
  return total_revenue - super
end
```

**Comment by CheezItMan** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Very good number of commits and good commit messages
Answer comprehension questions | Check
**Wave 1** |
Appropriate use of Ruby's `Time`| Check
`Trip` has a helper method to calculate duration| Check
`User` (passenger) has a method to calculate total cost of all trips| Check
Tests for wave 1 | 
**Wave 2** |
`Driver` inherits from `User` | Check
`Driver` has `add_driven_trip` method | Check
`Driver` has method to calculate average rating | Check
`Driver` has method to calculate net expenditures and it uses `super` | **MISSING**
`Driver` has a method to calculate total revenue | Check
Tests for wave 2 | Mostly working, see my inline notes.  No tests for `net_expenditures`
**Wave 3** |
`TripDispatcher` has a new method to create trips | **MISING**
creating a trip in `TripDispatcher` relies on methods in `Driver` and `User` (passenger) to modify their own attributes | **MISING**
Complex logic was correctly implemented | **MISING**
Tests for `request_trip` | **MISING**
Methods from wave 1 and 2 handle incomplete trips | **MISING**
Tests for wave 1 and 2 methods with incomplete trips | **MISING**
**Overall** | You got through most of wave 2, but didn't get to the `request_trip` or overriding the total_expenditures method.  See my inline notes, some of which deal with changes you would need for wave 3.
